### PR TITLE
rpc/jsonrpc: serve the cached chain config without opening a read transaction

### PR DESCRIPTION
**Pull request by AskAlexSharov**
problem: `eth_chainId`, `erigon_forks` and GraphQL `chainID` open an MDBX read transaction on every call only to read the chain config, which is cached after the first request; each `BeginTxn` takes the reader-slot lock.

Solution: `tryChainConfig` / `tryChainConfigWithGenesis` return the cached values before opening a transaction.

`eth_chainId` on n0, 12 cores, 2 alternating rounds: c16 186k -> 199k rps (+6.9%), CPU per request 52.7 -> 47.0 us (-11%); c64 is loader-capped at 215k rps, CPU per request 50.6 -> 45.3 us (-10%).

### CODE REVIEW

File: rpc/jsonrpc/eth_block.go [hunk 1]
```diff
@@ -260,6 +260,11 @@ func (api *APIImpl) GetBlockByHash(ctx context.Context, numberOrHash rpc.BlockNu
 	}
 
 	hash := *numberOrHash.BlockHash
+	if !numberOrHash.RequireCanonical && api.blocksLRU != nil {
+		if block, ok := api.blocksLRU.Get(hash); ok && block != nil {
+			return ethapi.RPCMarshalBlock(block, true, fullTx), nil
+		}
+	}
```
Comment (awskii) [inline rpc/jsonrpc/eth_block.go]: **eth_getBlockByHash cache hit returns before the header-existence and prune checks, so it can serve blocks the old path rejected**

The new early return reads `blocksLRU` before any tx is opened. It skips both gates that the tx path runs first: `rpchelper.GetBlockNumber` (HeaderNumber lookup, line 274, returns null when the hash is unknown) and `api.BaseAPI.checkPruneBlocks` (line 279, returns `state.PrunedError`). The early return runs whenever `RequireCanonical` is false, and a bare hash argument never sets it, so almost every real request takes this path. `blocksLRU` has no invalidation: the only `Add` is in `blockWithSenders` (eth_api.go:352-358), which many RPC methods call, and there is no `Remove`/`Purge`. Entries leave only under LRU pressure (128 entries, or 640 on a standalone rpcdaemon).

Prune case: with `--prune.distance.blocks=N`, block B is read while inside the window and gets cached. The head then moves past B+N while B stays in the LRU. `checkPruneField` (eth_api.go:744-749) compares against `Distance.PruneTo(latest)`, which moves with the head. Before this change `eth_getBlockByHash(B)` returned `PrunedError`; now it returns the full block, while `eth_getBlockByNumber(B)` still returns the error. The result depends on per-process cache state and changes after a restart. This also breaks the contract in `prune_gating_test.go:115`, which marks `eth_getBlockByHash` as `gatedByBlocks`. That test cannot catch it because `newPruneGatingAPIs` builds APIs with an empty LRU.

Visibility case: `InsertBlocks` flushes more than 16 blocks to the DB, B is read and cached, then `ValidateChain` finds the chain invalid and `purgeBadChain` calls `rawdb.DeleteHeader` (execution/execmodule/exec_module.go:652-716). This removes B's `HeaderNumber` entry. The old path returned null; the new path serves the invalid block until eviction. A similar case: an overlay-only block is cached while its overlay is published, then `PublishOverlay(nil)` discards it after a notification or commit failure. The old path returned not found, but the cached block is still served.

The new `TestGetBlockByHashServesCachedBlockWithoutReadTx` only checks that no tx is opened.

Fix: Keep cache hits behind the existing visibility and prune validation: either do the LRU lookup only after GetBlockNumber and checkPruneBlocks, or check the cached block's NumberU64() against a prune boundary that needs no tx (or bypass only when Blocks pruning is disabled). Also invalidate LRU entries when purgeBadChain deletes headers or an overlay is unpublished. Add a regression test that fills blocksLRU and then moves the prune boundary past the block or deletes its header.

File: rpc/jsonrpc/eth_api_test.go [hunk 2]
```diff
@@ -531,3 +533,65 @@ func TestStateMethods_OmittedBlockDefaultsToLatest(t *testing.T) {
 	a.NoError(err)
 	a.Equal(svLatest, svNil)
 }
+
+type noReadTxDB struct{ kv.TemporalRoDB }
+
+func (noReadTxDB) BeginTemporalRo(context.Context) (kv.TemporalTx, error) {
+	return nil, errors.New("unexpected read transaction")
+}
```
Comment (awskii) [inline rpc/jsonrpc/eth_api_test.go]: **noReadTxDB duplicates the existing unopenableDB test double in the same package**

The new `noReadTxDB` embeds `kv.TemporalRoDB` and overrides only `BeginTemporalRo` to return an error. `unopenableDB` in `rpc/jsonrpc/overlay_race_test.go:2503-2509` is the same type in the same `jsonrpc` test package: it embeds `kv.TemporalRoDB` and its `BeginTemporalRo` returns `errors.New("no read transaction available")`. The four new tests can use `unopenableDB{m.DB}` with no change in behavior. As it stands, the package has two hand-written stubs for one behavior, and they can drift apart.

Fix: Delete `noReadTxDB` and use `unopenableDB{TemporalRoDB: m.DB}` in the four new tests, then drop the `errors` import if nothing else uses it.

**Review by awskii — CHANGES_REQUESTED** [review]
Requesting changes because the `eth_getBlockByHash` cache hit returns before `GetBlockNumber` and `checkPruneBlocks`, so it can serve pruned or deleted blocks the transaction path rejects. One minor point inline.

File: rpc/jsonrpc/eth_block.go (hunk 1 above)
Comment (AskAlexSharov) [reply rpc/jsonrpc/eth_block.go]: Agreed, removed in 75e708a77a. A cache hit cannot run the header and prune checks without the read transaction, and it was flat on the fixture anyway. The PR now covers only the cached chain config.

File: rpc/jsonrpc/eth_api_test.go (hunk 2 above)
Comment (AskAlexSharov) [reply rpc/jsonrpc/eth_api_test.go]: Fixed in 75e708a77a: the tests use `unopenableDB`.

**Review by awskii — APPROVED** [review]
`75e708a` removes the `eth_getBlockByHash` cache hit, so the block path again runs its header and prune checks, and the tests use `unopenableDB`. One minor point, on the PR title, so it has no diff line:

**PR title still says it serves cached blocks without a read transaction, but that path was removed**

The live PR title is still `rpc/jsonrpc: serve cached chain config and blocks without opening a read transaction`. Commit 75e708a77a3 removed the `eth_getBlockByHash` blocksLRU early return and its test. At HEAD (b370029ddba), `GetBlockByHash` (rpc/jsonrpc/eth_block.go:262-279) again opens the read tx and runs `GetBlockNumber` and `checkPruneBlocks` before it reads the block. The PR now only caches the chain config for `eth_chainId`, `erigon_forks` and GraphQL `chainID` (erigon_system.go, eth_api.go, eth_api_test.go, eth_system.go, graphql_api.go). The body was updated to match, but the title was not. On squash-merge the title becomes the commit subject on main. History would then say eth_getBlockByHash serves cached blocks without a read tx, which is the exact behaviour that was rejected for skipping the prune and header-existence checks. That misleads anyone who later reads the history, bisects a prune-gating regression, writes a backport or prepares release notes. The problem is in PR metadata only; there is no runtime impact.

Fix: Rename the PR to `rpc/jsonrpc: serve cached chain config without opening a read transaction`.
